contracts/poap/src/msg.rs

```diff
@@ -0,0 +1,398 @@
+            },
+        };
+
+        assert_eq!(
+            ContractError::InvalidPerAddressLimit {},
+            msg.validate().unwrap_err()
+        );
+    }
+
+    #[test]
+    fn instantiate_with_invalid_event_uri_error() {
+        // Invalid uri
+        let msg = InstantiateMsg {
+            admin: "".to_string(),
+            minter: "".to_string(),
+            cw721_code_id: 0u64.into(),
+            cw721_initiate_msg: Cw721InstantiateMsg {
+                name: "".to_string(),
+                minter: "".to_string(),
+                symbol: "".to_string(),
+            },
+            event_info: EventInfo {
+                creator: "".to_string(),
+                start_time: Timestamp::from_seconds(1),
+                end_time: Timestamp::from_seconds(2),
+                per_address_limit: 1,
+                base_poap_uri: "ipfs://domain.com".to_string(),
+                event_uri: "invalid_event_uri".to_string(),
+            },
+        };
+
+        assert_eq!(
+            ContractError::InvalidEventUri {},
+            msg.validate().unwrap_err()
+        );
+
+        // Non ipfs uri
+        let msg = InstantiateMsg {
+            admin: "".to_string(),
+            minter: "".to_string(),
+            cw721_code_id: 0u64.into(),
+            cw721_initiate_msg: Cw721InstantiateMsg {
+                name: "".to_string(),
+                minter: "".to_string(),
+                symbol: "".to_string(),
+            },
+            event_info: EventInfo {
+                creator: "".to_string(),
+                start_time: Timestamp::from_seconds(1),
+                end_time: Timestamp::from_seconds(2),
+                per_address_limit: 1,
+                base_poap_uri: "ipfs://domain.com".to_string(),
+                event_uri: "https://domain.com".to_string(),
+            },
+        };
+
+        assert_eq!(
+            ContractError::InvalidEventUri {},
+            msg.validate().unwrap_err()
+        );
+    }
+
+    #[test]
+    fn instantiate_with_invalid_base_poap_uri_error() {
+        // Invalid uri
+        let msg = InstantiateMsg {
+            admin: "".to_string(),
+            minter: "".to_string(),
+            cw721_code_id: 0u64.into(),
+            cw721_initiate_msg: Cw721InstantiateMsg {
+                name: "".to_string(),
+                minter: "".to_string(),
+                symbol: "".to_string(),
+            },
+            event_info: EventInfo {
+                creator: "".to_string(),
+                start_time: Timestamp::from_seconds(1),
+                end_time: Timestamp::from_seconds(2),
+                per_address_limit: 1,
+                base_poap_uri: "https://domain.com".to_string(),
+                event_uri: "ipfs://domain.com".to_string(),
+            },
+        };
+
+        assert_eq!(
+            ContractError::InvalidPoapUri {},
+            msg.validate().unwrap_err()
+        );
+
+        // Non ipfs uri
```

Comment:
I think it can be splitted to `instantiate_with_invalid_base_poap_uri_non_ipfs_error`.